File: backend/Services/MessagingService.cs
using Microsoft.EntityFrameworkCore;
using SapBasisPulse.Api.Data;
using SapBasisPulse.Api.DTOs;
using SapBasisPulse.Api.Entities;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace SapBasisPulse.Api.Services
{
    public class MessagingService : IMessagingService
    {
        private readonly AppDbContext _context;

        public MessagingService(AppDbContext context)
        {
            _context = context;
        }

        public async Task<ConversationDto> CreateConversationAsync(CreateConversationDto dto, Guid userId)
        {
            // Get the order and verify user access
            var order = await _context.Orders
                .Include(o => o.CreatedByUser)
                .Include(o => o.Consultant)
                .FirstOrDefaultAsync(o => o.Id == dto.OrderId);

            if (order == null)
                throw new ArgumentException("Order not found");

            // Verify user has access to this order
            if (order.CreatedByUserId != userId && order.ConsultantId != userId)
                throw new UnauthorizedAccessException("User does not have access to this order");

            // Check if conversation already exists for this order
            var existingConversation = await _context.Conversations
                .FirstOrDefaultAsync(c => c.OrderId == dto.OrderId);

            if (existingConversation != null)
            {
                return await MapToConversationDto(existingConversation, userId);
            }

            // Create new conversation
            var conversation = new Conversation
            {
                Id = Guid.NewGuid(),
                OrderId = dto.OrderId,
                CustomerId = order.CreatedByUserId,
                ConsultantId = order.ConsultantId,
                Subject = dto.Subject,
                CreatedAt = DateTime.UtcNow,
                LastMessageAt = DateTime.UtcNow,
                IsActive = true
            };

            _context.Conversations.Add(conversation);

            // Create initial message
            var initialMessage = new Message
            {
                Id = Guid.NewGuid(),
                ConversationId = conversation.Id,
                SenderId = userId,
                Content = dto.InitialMessage,
                MessageType = "text",
                SentAt = DateTime.UtcNow
            };

            _context.Messages.Add(initialMessage);
            await _context.SaveChangesAsync();

            return await MapToConversationDto(conversation, userId);
        }

        public async Task<ConversationDto?> GetConversationByIdAsync(Guid conversationId, Guid userId)
        {
            var conversation = await _context.Conversations
                .Include(c => c.Customer)
                .Include(c => c.Consultant)
                .Include(c => c.Order)
                .FirstOrDefaultAsync(c => c.Id == conversationId);

            if (conversation == null || !await CanUserAccessConversationAsync(conversationId, userId))
                return null;

            return await MapToConversationDto(conversation, userId);
        }

        public async Task<ConversationDto?> GetConversationByOrderIdAsync(Guid orderId, Guid userId)
        {
            var conversation = await _context.Conversations
                .Include(c => c.Customer)
                .Include(c => c.Consultant)
                .Include(c => c.Order)
                .FirstOrDefaultAsync(c => c.OrderId == orderId);

            if (conversation == null || !await CanUserAccessConversationAsync(conversation.Id, userId))
                return null;

            return await MapToConversationDto(conversation, userId);
        }

        public async Task<IEnumerable<ConversationDto>> GetUserConversationsAsync(Guid userId, int page = 1, int pageSize = 20)
        {
            var conversations = await _context.Conversations
                .Include(c => c.Customer)
                .Include(c => c.Consultant)
                .Include(c => c.Order)
                .Where(c => c.CustomerId == userId || c.ConsultantId == userId)
                .OrderByDescending(c => c.LastMessageAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            var conversationDtos = new List<ConversationDto>();
            foreach (var conversation in conversations)
            {
                conversationDtos.Add(await MapToConversationDto(conversation, userId));
            }

            return conversationDtos;
        }

        public async Task<bool> UpdateConversationStatusAsync(Guid conversationId, bool isActive, Guid userId)
        {
            var conversation = await _context.Conversations
                .FirstOrDefaultAsync(c => c.Id == conversationId);

            if (conversation == null || !await CanUserAccessConversationAsync(conversationId, userId))
                return false;

            conversation.IsActive = isActive;
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<MessageDto> SendMessageAsync(CreateMessageDto dto, Guid senderId)
        {
            if (!await CanUserAccessConversationAsync(dto.ConversationId, senderId))
                throw new UnauthorizedAccessException("User does not have access to this conversation");

            var message = new Message
            {
                Id = Guid.NewGuid(),
                ConversationId = dto.ConversationId,
                SenderId = senderId,
                Content = dto.Content,
                MessageType = dto.MessageType,
                SentAt = DateTime.UtcNow
            };

            _context.Messages.Add(message);

            // Update conversation's last message time
            var conversation = await _context.Conversations
                .FirstOrDefaultAsync(c => c.Id == dto.ConversationId);
            
            if (conversation != null)
            {
                conversation.LastMessageAt = DateTime.UtcNow;
            }

            await _context.SaveChangesAsync();

            return await MapToMessageDto(message);
        }

        public async Task<MessageDto?> GetMessageByIdAsync(Guid messageId, Guid userId)
        {
            var message = await _context.Messages
                .Include(m => m.Conversation)
                .Include(m => m.Sender)
                .Include(m => m.Attachments)
                    .ThenInclude(a => a.UploadedBy)
                .FirstOrDefaultAsync(m => m.Id == messageId);

            if (message == null || !await CanUserAccessConversationAsync(message.ConversationId, userId))
                return null;

            return await MapToMessageDto(message);
        }

        public async Task<IEnumerable<MessageDto>> GetConversationMessagesAsync(Guid conversationId, Guid userId, int page = 1, int pageSize = 50)
        {
            if (!await CanUserAccessConversationAsync(conversationId, userId))
                throw new UnauthorizedAccessException("User does not have access to this conversation");

            var messages = await _context.Messages
                .Include(m => m.Sender)
                .Include(m => m.Attachments)
                    .ThenInclude(a => a.UploadedBy)
                .Where(m => m.ConversationId == conversationId)
                .OrderByDescending(m => m.SentAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            var messageDtos = new List<MessageDto>();
            foreach (var message in messages)
            {
                messageDtos.Add(await MapToMessageDto(message));
            }

            return messageDtos.OrderBy(m => m.SentAt);
        }

        public async Task<MessageDto?> UpdateMessageAsync(Guid messageId, UpdateMessageDto dto, Guid userId)
        {
            var message = await _context.Messages
                .Include(m => m.Sender)
                .Include(m => m.Attachments)
                    .ThenInclude(a => a.UploadedBy)
                .FirstOrDefaultAsync(m => m.Id == messageId);

            if (message == null || message.SenderId != userId)
                return null;

            message.Content = dto.Content;
            message.IsEdited = true;
            message.EditedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return await MapToMessageDto(message);
        }

        public async Task<bool> DeleteMessageAsync(Guid messageId, Guid userId)
        {
            var message = await _context.Messages
                .FirstOrDefaultAsync(m => m.Id == messageId);

            if (message == null || message.SenderId != userId)
                return false;

            _context.Messages.Remove(message);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> MarkMessagesAsReadAsync(MarkMessageReadDto dto, Guid userId)
        {
            var messages = await _context.Messages
                .Include(m => m.Conversation)
                .Where(m => dto.MessageIds.Contains(m.Id))
                .ToListAsync();

            var updated = 0;
            foreach (var message in messages)
            {
                // Only mark as read if user has access and message is not from the same user
                if (message.SenderId != userId && 
                    await CanUserAccessConversationAsync(message.ConversationId, userId))
                {
                    if (message.ReadAt == null)
                    {
                        message.ReadAt = DateTime.UtcNow;
                        updated++;
                    }
                }
            }

            if (updated > 0)
            {
                await _context.SaveChangesAsync();
            }

            return updated > 0;
        }

        public async Task<MessageAttachmentDto> AddAttachmentToMessageAsync(Guid messageId, string fileName, string originalFileName, string contentType, long fileSize, string filePath, Guid userId)
        {
            var message = await _context.Messages
                .Include(m => m.Conversation)
                .FirstOrDefaultAsync(m => m.Id == messageId);

            if (message == null || !await CanUserAccessConversationAsync(message.ConversationId, userId))
                throw new UnauthorizedAccessException("User does not have access to this message");

            var attachment = new MessageAttachment
            {
                Id = Guid.NewGuid(),
                MessageId = messageId,
                FileName = fileName,
                OriginalFileName = originalFileName,
                ContentType = contentType,
                FileSize = fileSize,
                FilePath = filePath,
                UploadedAt = DateTime.UtcNow,
                UploadedByUserId = userId
            };

            _context.MessageAttachments.Add(attachment);
            await _context.SaveChangesAsync();

            return await MapToAttachmentDto(attachment);
        }

        public async Task<MessageAttachmentDto?> GetAttachmentAsync(Guid attachmentId, Guid userId)
        {
            var attachment = await _context.MessageAttachments
                .Include(a => a.Message)
                    .ThenInclude(m => m.Conversation)
                .Include(a => a.UploadedBy)
                .FirstOrDefaultAsync(a => a.Id == attachmentId);

            if (attachment == null || !await CanUserAccessConversationAsync(attachment.Message.ConversationId, userId))
                return null;

            return await MapToAttachmentDto(attachment);
        }

        public async Task<bool> DeleteAttachmentAsync(Guid attachmentId, Guid userId)
        {
            var attachment = await _context.MessageAttachments
                .Include(a => a.Message)
                    .ThenInclude(m => m.Conversation)
                .FirstOrDefaultAsync(a => a.Id == attachmentId);

            if (attachment == null || 
                (attachment.UploadedByUserId != userId && !await CanUserAccessConversationAsync(attachment.Message.ConversationId, userId)))
                return false;

            _context.MessageAttachments.Remove(attachment);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<int> GetUnreadMessageCountAsync(Guid userId)
        {
            return await _context.Messages
                .Include(m => m.Conversation)
                .Where(m => (m.Conversation.CustomerId == userId || m.Conversation.ConsultantId == userId) &&
                           m.SenderId != userId &&
                           m.ReadAt == null)
                .CountAsync();
        }

        public async Task<bool> CanUserAccessConversationAsync(Guid conversationId, Guid userId)
        {
            return await _context.Conversations
                .AnyAsync(c => c.Id == conversationId && 
                              (c.CustomerId == userId || c.ConsultantId == userId));
        }

        private async Task<ConversationDto> MapToConversationDto(Conversation conversation, Guid userId)
        {
            var lastMessage = await _context.Messages
                .Include(m => m.Sender)
                .Where(m => m.ConversationId == conversation.Id)
                .OrderByDescending(m => m.SentAt)
                .FirstOrDefaultAsync();

            var unreadCount = await _context.Messages
                .Where(m => m.ConversationId == conversation.Id && 
                           m.SenderId != userId && 
                           m.ReadAt == null)
                .CountAsync();

            return new ConversationDto
            {
                Id = conversation.Id,
                OrderId = conversation.OrderId,
                OrderNumber = conversation.Order?.OrderNumber ?? "",
                CustomerId = conversation.CustomerId,
                CustomerName = $"{conversation.Customer?.FirstName} {conversation.Customer?.LastName}".Trim(),
                ConsultantId = conversation.ConsultantId,
                ConsultantName = conversation.Consultant != null ? 
                    $"{conversation.Consultant.FirstName} {conversation.Consultant.LastName}".Trim() : "",
                Subject = conversation.Subject,
                CreatedAt = conversation.CreatedAt,
                LastMessageAt = conversation.LastMessageAt,
                IsActive = conversation.IsActive,
                UnreadCount = unreadCount,
                LastMessage = lastMessage != null ? await MapToMessageDto(lastMessage) : null
            };
        }

        private async Task<MessageDto> MapToMessageDto(Message message)
        {
            var attachments = new List<MessageAttachmentDto>();
            if (message.Attachments?.Any() == true)
            {
                foreach (var attachment in message.Attachments)
                {
                    attachments.Add(await MapToAttachmentDto(attachment));
                }
            }

            return new MessageDto
            {
                Id = message.Id,
                ConversationId = message.ConversationId,
                SenderId = message.SenderId,
                SenderName = $"{message.Sender?.FirstName} {message.Sender?.LastName}".Trim(),
                SenderRole = message.Sender?.Role.ToString() ?? "Unknown",
                Content = message.Content,
                MessageType = message.MessageType,
                SentAt = message.SentAt,
                ReadAt = message.ReadAt,
                IsEdited = message.IsEdited,
                EditedAt = message.EditedAt,
                Attachments = attachments
            };
        }

        private async Task<MessageAttachmentDto> MapToAttachmentDto(MessageAttachment attachment)
        {
            return await Task.FromResult(new MessageAttachmentDto
            {
                Id = attachment.Id,
                FileName = attachment.FileName,
                OriginalFileName = attachment.OriginalFileName,
                ContentType = attachment.ContentType,
                FileSize = attachment.FileSize,
                UploadedAt = attachment.UploadedAt,
                UploadedByName = $"{attachment.UploadedBy?.FirstName} {attachment.UploadedBy?.LastName}".Trim(),
                DownloadUrl = $"/api/Messaging/attachments/{attachment.Id}/download"
            });
        }
    }
}

File: frontend/src/services/messagingService.ts
import axios from 'axios';

// Types for the messaging system
export interface Conversation {
  id: string;
  orderId: string;
  customerId: string;
  consultantId: string;
  isActive: boolean;
  createdAt: string;
  lastMessageAt: string;
  unreadCount: number;
  customerName: string;
  consultantName: string;
  orderNumber: string;
  subject: string;
}

export interface Message {
  id: string;
  conversationId: string;
  senderId: string;
  senderName: string;
  senderRole: string;
  content: string;
  messageType: 'text' | 'file' | 'system';
  isRead: boolean;
  isEdited: boolean;
  sentAt: string;
  readAt?: string;
  editedAt?: string;
  attachments: MessageAttachment[];
}

export interface MessageAttachment {
  id: string;
  messageId?: string;
  fileName: string;
  originalFileName: string;
  contentType: string;
  fileSize: number;
  uploadedAt: string;
  uploadedByName: string;
  downloadUrl: string;
}

// DTOs to match backend
export interface CreateConversationDto {
  orderId: string;
  subject: string;
  initialMessage: string;
}

export interface CreateMessageDto {
  conversationId: string;
  content: string;
  messageType: 'text' | 'file' | 'system';
}

export interface UpdateMessageDto {
  content: string;
}

export interface MarkMessageReadDto {
  messageIds: string[];
}

export interface PaginatedResult<T> {
  items: T[];
  currentPage: number;
  pageCount: number;
  pageSize: number;
  totalCount: number;
}

export interface FileUploadInfo {
  allowedFileTypes: string[];
  maxFileSize: number;
  maxFileSizeMB: number;
}

const API_BASE = import.meta.env.VITE_API_BASE || 'http://localhost:5274/api/';
const API_URL = `${API_BASE}Messaging`;

// Configure axios to include auth token
const getAuthHeaders = () => {
  const token = localStorage.getItem('authToken');
  return token ? { 'Authorization': `Bearer ${token}` } : {};
};

const MessagingService = {
  // Conversation endpoints
  createConversation: async (data: CreateConversationDto): Promise<Conversation> => {
    const response = await axios.post<Conversation>(`${API_URL}/conversations`, data, {
      headers: getAuthHeaders()
    });
    return response.data;
  },

  getUserConversations: async (page: number = 1, pageSize: number = 20): Promise<Conversation[]> => {
    const response = await axios.get<Conversation[]>(
      `${API_URL}/conversations?page=${page}&pageSize=${pageSize}`,
      { headers: getAuthHeaders() }
    );
    return response.data;
  },

  getConversation: async (conversationId: string): Promise<Conversation> => {
    const response = await axios.get<Conversation>(`${API_URL}/conversations/${conversationId}`, {
      headers: getAuthHeaders()
    });
    return response.data;
  },

  getConversationByOrder: async (orderId: string): Promise<Conversation> => {
    const response = await axios.get<Conversation>(`${API_URL}/conversations/by-order/${orderId}`, {
      headers: getAuthHeaders()
    });
    return response.data;
  },

  createOrGetConversationForOrder: async (orderId: string): Promise<Conversation> => {
    try {
      const response = await axios.post<Conversation>(`${API_URL}/conversations/for-order/${orderId}`, {}, {
        headers: getAuthHeaders()
      });
      return response.data;
    } catch (error: any) {
      console.error(`Error creating conversation for order ${orderId}:`, error.response?.data || error.message);
      throw error; // Re-throw for handling in the component
    }
  },

  updateConversationStatus: async (conversationId: string, isActive: boolean): Promise<boolean> => {
    const response = await axios.put<{ success: boolean }>(
      `${API_URL}/conversations/${conversationId}/status`, 
      isActive
    );
    return response.data.success;
  },

  // Message endpoints
  sendMessage: async (dto: CreateMessageDto): Promise<Message> => {
    const response = await axios.post<Message>(`${API_URL}/messages`, dto, {
      headers: getAuthHeaders()
    });
    return response.data;
  },

  getConversationMessages: async (
    conversationId: string, 
    page: number = 1, 
    pageSize: number = 50
  ): Promise<Message[]> => {
    const response = await axios.get<Message[]>(
      `${API_URL}/conversations/${conversationId}/messages?page=${page}&pageSize=${pageSize}`,
      { headers: getAuthHeaders() }
    );
    return response.data;
  },

  getMessage: async (messageId: string): Promise<Message> => {
    const response = await axios.get<Message>(`${API_URL}/messages/${messageId}`, {
      headers: getAuthHeaders()
    });
    return response.data;
  },

  updateMessage: async (messageId: string, content: string): Promise<Message> => {
    const response = await axios.put<Message>(
      `${API_URL}/messages/${messageId}`, 
      { content } as UpdateMessageDto,
      { headers: getAuthHeaders() }
    );
    return response.data;
  },

  deleteMessage: async (messageId: string): Promise<boolean> => {
    const response = await axios.delete<{ success: boolean }>(`${API_URL}/messages/${messageId}`, {
      headers: getAuthHeaders()
    });
    return response.data.success;
  },

  markMessagesAsRead: async (messageIds: string[]): Promise<boolean> => {
    const response = await axios.post<{ success: boolean }>(
      `${API_URL}/messages/mark-read`, 
      { messageIds } as MarkMessageReadDto,
      { headers: getAuthHeaders() }
    );
    return response.data.success;
  },

  // Attachment endpoints
  uploadAttachment: async (messageId: string, file: File): Promise<MessageAttachment> => {
    const formData = new FormData();
    formData.append('file', file);
    
    const response = await axios.post<MessageAttachment>(
      `${API_URL}/messages/${messageId}/attachments`, 
      formData,
      {
        headers: {
          ...getAuthHeaders(),
          'Content-Type': 'multipart/form-data',
        },
      }
    );
    
    return response.data;
  },

  getAttachment: async (attachmentId: string): Promise<MessageAttachment> => {
    const response = await axios.get<MessageAttachment>(`${API_URL}/attachments/${attachmentId}`, {
      headers: getAuthHeaders()
    });
    return response.data;
  },

  getAttachmentDownloadUrl: (attachmentId: string): string => {
    return `${API_URL}/attachments/${attachmentId}/download`;
  },

  deleteAttachment: async (attachmentId: string): Promise<boolean> => {
    const response = await axios.delete<{ success: boolean }>(`${API_URL}/attachments/${attachmentId}`, {
      headers: getAuthHeaders()
    });
    return response.data.success;
  },

  // Utility endpoints
  getUnreadMessageCount: async (): Promise<number> => {
    try {
      const response = await axios.get<{ unreadCount: number }>(`${API_URL}/unread-count`, {
        headers: getAuthHeaders()
      });
      return response.data.unreadCount || 0;
    } catch (error) {
      console.warn('Failed to fetch unread message count:', error);
      return 0;
    }
  },

  getFileUploadInfo: async (): Promise<FileUploadInfo> => {
    const response = await axios.get<FileUploadInfo>(`${API_URL}/file-upload-info`, {
      headers: getAuthHeaders()
    });
    return response.data;
  },
};

export default MessagingService;